delegate-app/Main.hs

```diff
@@ -5,45 +5,60 @@ import Prelude
 
 -- Haskell imports
 
-import Control.Concurrent (threadDelay)
+import Control.Concurrent (Chan, newChan, readChan, threadDelay, writeChan)
+import Control.Concurrent.Async (async)
+import Control.Monad (forever, void)
 import Control.Monad.IO.Class (MonadIO (liftIO))
 
+-- Plutus imports
+import Plutus.V2.Ledger.Api (POSIXTime (..))
+
 -- HydraAuction imports
+
 import HydraAuction.Delegate (
-  DelegateEvent (Start),
+  DelegateEvent (..),
   DelegateInput (DelegateEvent),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
+import HydraAuction.Delegate.Interface (DelegateResponse (..))
+import HydraAuction.OnChain.Common (secondsLeftInInterval, stageToInterval)
+import HydraAuction.Tx.Common (currentAuctionStage, currentTimeMilliseconds)
+
+consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
+consumer events = do
+  -- This will block until an event appears in the queue
+  ev <- liftIO $ readChan events
+  delegateResponse <- delegateStep ev
+  liftIO $ mbQueueAuctionPhases delegateResponse events
+  -- FIXME: send response to client
+  liftIO $
+    putStrLn $
+      "Delegate responses for input: " <> show delegateResponse
 
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
-eventsProducer thisIsFirstRun =
-  if thisIsFirstRun
-    then return $ Just Start
-    else return Nothing
-
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
-
-delegateTick :: Int
-delegateTick = 1_000
+mbQueueAuctionPhases :: [DelegateResponse] -> Chan DelegateInput -> IO ()
+mbQueueAuctionPhases [AuctionSet terms] events = queueCurrentStage
+  where
+    queueCurrentStage = do
+      currentStage <- currentAuctionStage terms
+      writeChan events (DelegateEvent $ AuctionStageStarted currentStage)
+      currentPosixTime <- POSIXTime <$> currentTimeMilliseconds
+      let mSecsLeft = secondsLeftInInterval currentPosixTime (stageToInterval terms currentStage)
+      case mSecsLeft of
+        Nothing -> pure ()
+        Just s -> do
+          void $ async (threadDelay (fromInteger s * 1000) >> queueCurrentStage)
```

Comment:
I do not understand async here. Should not this just lead to infinite nested thread creation?

I think just the single async with infinite loop should do the job, like in `consumer`:

```
-- This could only happen once per auction
case delegateResponse of
   [AuctionSet terms]) ->  liftIO $ queueCurrentStages delegateInputsQueue
   _ -> pure ()
```

Reply:
Eventually `mSecsLeft` will be `Nothing` (when we reach the cleanup phase, as it doesn't have an upper bound).

Reply:
Yes. What should it break? Thread just will not recurse then and thread gonna stop.